codex/sales/salesagent.nim

```diff
@@ -0,0 +1,85 @@
+import pkg/chronos
+import pkg/upraises
+import pkg/stint
+import ./statemachine
+import ./states/[downloading, unknown]
+import ../contracts/requests
+import ../rng
+
+proc newSalesAgent*(sales: Sales,
+                    requestId: RequestId,
+                    availability: ?Availability,
+                    request: ?StorageRequest): SalesAgent =
+  SalesAgent(
+    sales: sales,
+    requestId: requestId,
+    availability: availability,
+    request: request)
+
+# fwd declarations
+proc subscribeCancellation*(agent: SalesAgent): Future[void] {.gcsafe.}
+proc subscribeFailure*(agent: SalesAgent): Future[void] {.gcsafe.}
+
+proc populateRequest(agent: SalesAgent) {.async.} =
```

Comment:
I would call this `retrieveRequest`, because the name `populateRequest` is already used elsewhere with a different meaning (populating the fields of a storage request object)

Reply:
Done